aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1284 @@
+                },
+                (None, DEPENDENCY_SAFE_FLAG) => {
+                    // May not update SAFE flag at a future incorrect time (i.e. ABA), as observing
+                    // num_stalls = 0 under status is required to set SAFE flag, but impossible
+                    // until the corresponding unstall (that starts only after this stall finishes).
+                    self.dependency_shortcut
+                        .store(DEPENDENCY_DEFER_FLAG, Ordering::Relaxed);
+                },
+                (None, DEPENDENCY_EXECUTING_FLAG | DEPENDENCY_DEFER_FLAG) => {
+                    // Executing or aborted: shortcut not affected.
+                },
+                (Some(_), DEPENDENCY_SAFE_FLAG | DEPENDENCY_EXECUTING_FLAG) => {
+                    return Err(code_invariant_error(
+                        "Inconsistent status and dependency shortcut in stall",
+                    ));
+                },
+                (_, unsupported_flag_value) => {
+                    return Err(code_invariant_error(format!(
+                        "Unsupported flag value {unsupported_flag_value} in stall",
+                    )));
+                },
+            }
+
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    // Returns true if this unstall call changed the state, i.e. decreased num_stalls to 0.
+    // If so, scheduling hook is also invoked on the proxy.
+    pub(crate) fn unstall(&self) -> Result<bool, PanicError> {
+        let prev_num_stalls = self.num_stalls.fetch_sub(1, Ordering::SeqCst);
+
+        if prev_num_stalls == 0 {
+            return Err(code_invariant_error(format!(
+                "unstall called when num_stalls == 0"
+            )));
+        }
+
+        if prev_num_stalls == 1 {
+            // Acquire write lock for (non-monitor) shortcut modifications.
+            let inner_status = self.inner_status.lock();
+
+            // num_stalls updates are not under the lock, so need to re-check (otherwise
+            // a different stall might have already incremented the count.
+            let not_stalled = self.num_stalls.load(Ordering::Relaxed) == 0;
+
+            if not_stalled {
+                if let Some(incarnation) = inner_status.requires_execution() {
+                    if incarnation == 0 {
+                        // Invariant due to scheduler logic: for a successful unstall there must
+                        // have been a stall for incarnation 0, which is impossible (see above).
+                        return Err(code_invariant_error(
+                            "0-th incarnation may not be unstalled",
+                        ));
+                    }
+                    self.scheduler_proxy
+                        .add_to_schedule(incarnation == 1, self.txn_idx);
+                } else if inner_status.is_executed() {
+                    defer! {
+                    self.notify_waiting_workers();
+                    }
```

Comment:
This is the case when after execution I am stalled and then unstalled, but in the meantime some txns are wating, right?